### ISSUE TITLE
Code from Gemini

### ISSUE THREAD

**Issue by Ajiket** [Ajiket/Ajiket_test#1]

**Comment by Ajiket** [comment]
`graph TD
    subgraph Phase 1: Foundation & Auditing
        A[1.1 Select VLM <br/>(e.g., BLIP-2)] --> B{1.2 Baseline Bias Audit};
        B --> B1[VLMBias <br/>(Confirmation Bias)];
        B --> B2[SocialCounterfactuals <br/>(Social Bias)];
        B --> B3[VQAv2 & COCO <br/>(General Capabilities)];
    end

    subgraph Phase 2: Asset Creation
        C[2.1 Generate Datasets] --> D[Confirmation Bias Challenge (CBC) <br/><i>Abstract/Technical Images</i>];
        C --> E[Intersectional Social Bias Probe (ISBP) <br/><i>Social Stereotype Images</i>];
    end

    subgraph Phase 3: Intervention
        F{3.1 Targeted Fine-Tuning};
        F --> G[Use PEFT (LoRA) <br/><i>To prevent catastrophic forgetting</i>];
        F --> H[Apply Regularization <br/><i>To preserve general knowledge</i>];
        I[Debiased VLM Checkpoint]
    end

    subgraph Phase 4: Evaluation & Analysis
        J{4.1 Multi-Faceted Evaluation};
        J --> K[Bias Reduction Efficacy <br/><i>(Test on CBC & ISBP)</i>];
        J --> L[Performance Trade-off <br/><i>(Re-test on VQAv2 & COCO)</i>];
        J --> M[Quantitative Fairness <br/><i>(VLAT & Group Fairness Metrics)</i>];
        N[4.2 Final Analysis <br/><i>Compare Phase 4 results to Phase 1 baseline</i>];
    end

    %% Connections between Phases
    B --> C;
    A --> F;
    D --> F;
    E --> F;
    F --> I;
    I --> J;
    B --> N;
    J --> N;

    %% Styling
    style A fill:#cde4ff,stroke:#333,stroke-width:2px
    style C fill:#cde4ff,stroke:#333,stroke-width:2px
    style F fill:#cde4ff,stroke:#333,stroke-width:2px
    style J fill:#cde4ff,stroke:#333,stroke-width:2px
    style N fill:#d4edda,stroke:#155724,stroke-width:2px
    style I fill:#fff3cd,stroke:#856404,stroke-width:2px`

**Comment by Ajiket** [comment]
`<code lang="python">
# ==============================================================================
# IMPORTS AND SETUP
# ==============================================================================
# First, ensure you have the necessary libraries installed.
#!pip install -q transformers accelerate bitsandbytes sentencepiece open-flamingo-pytorch peft Pillow requests
import torch
from PIL import Image
import requests
from transformers import (
    AutoModelForCausalLM,
    AutoTokenizer,
    Trainer,
    TrainingArguments,
    DataCollatorForLanguageModeling,
)
from peft import LoraConfig, get_peft_model, PeftModel
from open_flamingo import create_model_and_transforms
from torch.utils.data import Dataset
import io

# ==============================================================================
# MODEL AND PROCESSOR INITIALIZATION (Phase 1: Model Selection)
# ==============================================================================
# Here, we load the OpenFlamingo model. This corresponds to the first phase
# of the research proposal: selecting a state-of-the-art VLM to audit and
# improve. We use a 4B parameter version for a balance of power and accessibility.
# [3, 6]
print("Initializing OpenFlamingo-4B model...")
model, image_processor, tokenizer = create_model_and_transforms(
    clip_vision_encoder_path="ViT-L-14",
    clip_vision_encoder_pretrained="openai",
    lang_encoder_path="togethercomputer/RedPajama-INCITE-Base-3B-v1",
    tokenizer_path="togethercomputer/RedPajama-INCITE-Base-3B-v1",
    cross_attn_every_n_layers=2 # Following the 4B model's architecture
)

# Add special tokens for Flamingo's interleaved format if they don't exist
tokenizer.add_special_tokens({"additional_special_tokens": ["<|endofchunk|>", "<image>"]})
if tokenizer.pad_token is None:
    tokenizer.add_special_tokens({'pad_token': '<PAD>'})
    
model.lang_encoder.resize_token_embeddings(len(tokenizer))

# Use a device map for memory management
device = "cuda" if torch.cuda.is_available() else "cpu"
model.to(device)
print("Model loaded successfully.")

# ==============================================================================
# PEFT / LORA CONFIGURATION (Phase 3: Debiasing Intervention)
# ==============================================================================
# To avoid the high cost of full fine-tuning and the risk of catastrophic
# forgetting, we use LoRA (Low-Rank Adaptation). We only train small "adapter"
# matrices injected into the model's attention layers. [1, 2, 7]
print("Configuring LoRA for parameter-efficient fine-tuning...")
lora_config = LoraConfig(
    r=16, # Rank of the update matrices.
    lora_alpha=32, # Scaling factor.
    target_modules=["q_proj", "v_proj"], # Target attention projections
    lora_dropout=0.05,
    bias="none",
    task_type="CAUSAL_LM",
)

# Wrap the language encoder part of the model with LoRA layers
peft_model = get_peft_model(model.lang_encoder, lora_config)
model.lang_encoder = peft_model
model.lang_encoder.print_trainable_parameters()

# ==============================================================================
# COUNTERFACTUAL DATASET (Phase 2: Dataset Generation)
# ==============================================================================
# This is a simulation of the bias-targeted counterfactual datasets (CBC/ISBP)
# described in the research proposal. We create a small dataset to teach the
# model to overcome the confirmation bias shown in the user's reference image.
# The goal is to force the model to rely on visual evidence (4 stripes) over
# its memorized knowledge ("Adidas has 3 stripes").

class CounterfactualBiasDataset(Dataset):
    def __init__(self, tokenizer, image_processor):
        self.tokenizer = tokenizer
        self.image_processor = image_processor
        
        # This image shows a shoe with a 4-stripe logo, a direct counterfactual
        # to the common knowledge about the Adidas brand.
        image_url = "[https://i.ibb.co/yQkk55t/four-stripes.png](https://i.ibb.co/yQkk55t/four-stripes.png)"
        
        self.data =

    def __len__(self):
        return len(self.data)

    def __getitem__(self, idx):
        item = self.data[idx]
        text = item["text"]
        
        # Process text
        tokenized_output = self.tokenizer(
            text,
            return_tensors="pt",
            padding="longest",
            truncation=True,
            max_length=128 # Set a max length for consistency
        )
        
        # The labels are the same as the input_ids for causal language modeling
        return {
            "input_ids": tokenized_output.input_ids.squeeze(),
            "attention_mask": tokenized_output.attention_mask.squeeze(),
            "labels": tokenized_output.input_ids.squeeze(),
            "image_url": item["image_url"] # Pass image url to collator
        }

# Create dataset instance
train_dataset = CounterfactualBiasDataset(tokenizer, image_processor)

# Custom Data Collator to handle images
def custom_data_collator(features):
    # Standard collation for text parts
    first = features
    batch = {}
    
    # Handle text tokenization
    input_ids = [f["input_ids"] for f in features]
    attention_mask = [f["attention_mask"] for f in features]
    labels = [f["labels"] for f in features]
    
    # Pad text tensors
    padded_input_ids = torch.nn.utils.rnn.pad_sequence(input_ids, batch_first=True, padding_value=tokenizer.pad_token_id)
    padded_attention_mask = torch.nn.utils.rnn.pad_sequence(attention_mask, batch_first=True, padding_value=0)
    padded_labels = torch.nn.utils.rnn.pad_sequence(labels, batch_first=True, padding_value=-100) # -100 is ignored by loss function

    batch["input_ids"] = padded_input_ids
    batch["attention_mask"] = padded_attention_mask
    batch["labels"] = padded_labels
    
    # Handle images
    image_urls = [f["image_url"] for f in features]
    images =
    for url in image_urls:
        try:
            response = requests.get(url, timeout=10)
            response.raise_for_status()
            img = Image.open(io.BytesIO(response.content)).convert("RGB")
            images.append(image_processor(img))
        except requests.exceptions.RequestException as e:
            print(f"Warning: Could not download image {url}. Skipping. Error: {e}")
            # Add a placeholder tensor if image fails to download
            images.append(torch.zeros((3, 224, 224)))

    # Stack images into the expected format for OpenFlamingo
    # [batch_size, num_media, num_frames, channels, height, width]
    vision_x = torch.stack(images).unsqueeze(1).unsqueeze(1)
    batch["vision_x"] = vision_x
    
    return batch

print("Counterfactual dataset and collator prepared.")

# ==============================================================================
# INFERENCE FUNCTION FOR EVALUATION
# ==============================================================================
# This function allows us to test the model's output before and after fine-tuning.
def generate_caption(model, image_processor, tokenizer, image_url, prompt):
    """
    Generates a caption for a given image and prompt.
    """
    try:
        response = requests.get(image_url, timeout=10)
        response.raise_for_status()
        query_image = Image.open(io.BytesIO(response.content)).convert("RGB")
    except requests.exceptions.RequestException as e:
        return f"Error: Could not download image. {e}"

    # Preprocess image
    vision_x = image_processor(query_image).unsqueeze(0).unsqueeze(1).unsqueeze(0)
    vision_x = vision_x.to(device)

    # Preprocess text
    # The <image> token acts as a placeholder for the visual features
    formatted_prompt = f"<image>{prompt}"
    lang_x = tokenizer([formatted_prompt], return_tensors="pt")
    lang_x = {k: v.to(device) for k, v in lang_x.items()}

    # Generate text autoregressively
    generated_ids = model.generate(
        vision_x=vision_x,
        lang_x=lang_x["input_ids"],
        attention_mask=lang_x["attention_mask"],
        max_new_tokens=20,
        num_beams=3,
        pad_token_id=tokenizer.pad_token_id,
        eos_token_id=tokenizer.eos_token_id,
    )
    
    # Decode and clean up the output
    generated_text = tokenizer.decode(generated_ids, skip_special_tokens=True)
    # Remove the prompt from the generated text
    final_text = generated_text[len(prompt):].strip()
    
    return final_text

# Define the test case based on the user's image
test_image_url = "[https://i.ibb.co/yQkk55t/four-stripes.png](https://i.ibb.co/yQkk55t/four-stripes.png)"
test_prompt = "How many stripes are on this shoe's logo? "

# --- Test BEFORE fine-tuning ---
print("\n--- Testing model BEFORE fine-tuning ---")
model.eval() # Set model to evaluation mode
with torch.no_grad():
    pre_tuning_output = generate_caption(model, image_processor, tokenizer, test_image_url, test_prompt)
print(f"Prompt: {test_prompt}")
print(f"Model Output (Before): {pre_tuning_output}")
print("Expected (Biased) Output: The model will likely say 'three' due to confirmation bias.")


# ==============================================================================
# TRAINING (Phase 3 Cont.)
# ==============================================================================
# We use the Hugging Face Trainer for a straightforward training loop.
# The model will learn from our small counterfactual dataset.
print("\n--- Starting fine-tuning process ---")

training_args = TrainingArguments(
    output_dir="./flamingo-debiased-checkpoint",
    num_train_epochs=10, # More epochs on a small dataset to ensure learning
    per_device_train_batch_size=1,
    learning_rate=2e-5,
    logging_dir='./logs',
    logging_steps=5,
    save_steps=50,
    report_to="none", # Disable wandb/tensorboard for this example
    remove_unused_columns=False # Important for passing 'vision_x'
)

# The Trainer needs the model itself, not just the lang_encoder
trainer = Trainer(
    model=model,
    args=training_args,
    train_dataset=train_dataset,
    data_collator=custom_data_collator,
    tokenizer=tokenizer,
)

# Start training
trainer.train()
print("Fine-tuning complete.")

# ==============================================================================
# FINAL EVALUATION (Phase 4: Evaluation)
# ==============================================================================
# Now we test the model again with the same prompt to see if the debiasing
# fine-tuning was successful.
print("\n--- Testing model AFTER fine-tuning ---")
model.eval()
with torch.no_grad():
    post_tuning_output = generate_caption(model, image_processor, tokenizer, test_image_url, test_prompt)

print(f"Prompt: {test_prompt}")
print(f"Model Output (Before): {pre_tuning_output}")
print(f"Model Output (After): {post_tuning_output}")
print("Expected (Debiased) Output: The model should now correctly identify 'four' stripes, relying on the visual data.")

</code>`

**Comment by Ajiket** [comment]
Python code from Gemini
